tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,439 @@
+    """Import graph_defs.
+    
+    The current implementation loads graph_defs from memory."""
+    op_to_graph_def = {op: _get_graph_def(filepath) for op, filepath
+                       in op_to_filepath.items()}
+    return op_to_graph_def
+
+
+def _get_graph_def(filepath):
+    graph_def = graph_pb2.GraphDef()
+    with tf.compat.v1.gfile.FastGFile(filepath, 'rb') as f:
+        graph_def.ParseFromString(f.read())
+    return graph_def
+
+
+"""Partitioning"""
+def partition_all_graphs(op_to_graph_def, op_to_outputs):
+    """The main method to call."""
+    op_to_execution_specs = {}
+    for op in op_to_graph_def:
+        execution_specs = _partition_one_graph(op_to_graph_def[op],
+                                                  op_to_outputs[op])
+        op_to_execution_specs[op] = execution_specs
+    return op_to_execution_specs
+
+
+def _partition_one_graph(graph_def, outputs):
+    """Partition a graph_def.
+    
+    Arguments:
+        graph_def: a GraphDef proto
+        outputs: a set of output node names
+    
+    Variables:
+        graph: a tf.Graph instance
+        node_name_to_node_def: {node name: a NodeDef}
+        node_name_to_input_names: {node name: a list of input names}
+        remote_op_relations: {remote op name: a list of remote op children}
+    
+    Returns:
+        execution_specs: a list of specs, each spec contains:
+                         {'subgraph': a GraphDef,
+                          'inputs': a set of node names,
+                          'outputs': a set of node names,
+                          'is_remote_op': a Boolean,
+                          'body_nodes': not important,
+                          'nodes_from_other_layers': not important}
+    """
+    graph = _get_graph(graph_def)
+    node_name_to_node_def = _get_node_name_to_node_def(graph_def)
+    node_name_to_input_names = _get_node_name_to_input_names(graph_def)
+    
+    remote_op_relations = _get_remote_op_relations(graph_def,
+                                                   node_name_to_node_def,
+                                                   node_name_to_input_names)
+    
+    execution_specs = _get_execution_specs(graph_def,
+                                           graph,
+                                           node_name_to_node_def,
+                                           node_name_to_input_names,
+                                           remote_op_relations,
+                                           outputs)
+    # DEBUG
+    _debug_print_execution_specs(execution_specs)
+    
+    return execution_specs
+
+
+def _debug_print_execution_specs(execution_specs):
+    for execution_spec in execution_specs:
+        print('\nIs the current spec describing a remote op?', execution_spec['is_remote_op'])
+        print('Inputs:', execution_spec['inputs'])
+        print('Outputs:', execution_spec['outputs'])
+        print('Body nodes:', execution_spec['body_nodes'])
```

Comment:
Remember to remove this before committing.

Reply:
Removed. BTW, do users want to test? wouldn't this printing method be helpful?

Reply:
In that case, you should use `logging.debug` or `logging.info` so that the user can suppress the output.